File: swift-facade/Sources/DocxktDSL/LineNumbering.swift
import Docxkt

/// Document-level line numbering (`<w:lnNumType>`). Sits as a top-level
/// `Document { LineNumbering(...) }` block.
///
/// ```swift
/// Document {
///     LineNumbering(countBy: 1, start: 1, distance: 360, restart: .newPage)
///     Paragraph { Text("Numbered every line, restart per page") }
/// }
/// ```
public struct LineNumbering {
    public var countBy: Int?
    public var start: Int?
    public var distance: Int?
    public var restart: LineNumberRestartKind?

    public init(
        countBy: Int? = nil,
        start: Int? = nil,
        distance: Int? = nil,
        restart: LineNumberRestartKind? = nil,
    ) {
        self.countBy = countBy
        self.start = start
        self.distance = distance
        self.restart = restart
    }

    internal func applyToDocument(_ scope: KotlinDocumentScope) {
        scope.lineNumbering(
            countBy: countBy.map { KotlinInt(int: Int32($0)) },
            start: start.map { KotlinInt(int: Int32($0)) },
            distanceTwips: distance.map { KotlinInt(int: Int32($0)) },
            restart: restart?.kotlin,
        )
    }
}

File: core/src/commonMain/kotlin/io/docxkt/model/section/LineNumbering.kt
// Port of: src/file/document/body/section-properties/properties/line-number.ts
package io.docxkt.model.section

import io.docxkt.xml.XmlComponent
import io.docxkt.xml.selfClosingElement

/**
 * `<w:lnNumType w:restart="…"/>` value — when line numbering
 * resets to its starting value.
 */
public enum class LineNumberRestart(internal val wire: String) {
    /** Continue from the previous section's line numbering. */
    CONTINUOUS("continuous"),

    /** Restart line numbering at the top of each page. */
    NEW_PAGE("newPage"),

    /** Restart line numbering at the start of each section. */
    NEW_SECTION("newSection"),
}

/**
 * `<w:lnNumType>` — line numbering for a section.
 *
 * - [countBy] — display only multiples of this value
 *   (e.g. `5` shows every 5th line).
 * - [start] — starting value when the count restarts.
 * - [distance] — twips between text margin and line-number gutter.
 * - [restart] — when to reset numbering.
 *
 * All four attributes are optional. Pass-through; we don't
 * validate `start ≥ 1` or other range constraints.
 */
internal class LineNumbering(
    val countBy: Int? = null,
    val start: Int? = null,
    val distance: Int? = null,
    val restart: LineNumberRestart? = null,
) : XmlComponent("w:lnNumType") {

    override fun appendXml(out: Appendable) {
        out.selfClosingElement(
            "w:lnNumType",
            "w:countBy" to countBy?.toString(),
            "w:start" to start?.toString(),
            "w:restart" to restart?.wire,
            "w:distance" to distance?.toString(),
        )
    }
}
